src/meta-srv/src/election/postgres.rs

```diff
@@ -0,0 +1,488 @@
+
+use common_meta::kv_backend::postgres::{
+    CAS, POINT_DELETE, POINT_GET, PREFIX_SCAN, PUT_IF_NOT_EXISTS,
+};
+use serde::{Deserialize, Serialize};
+use snafu::{ensure, ResultExt};
+use tokio::sync::broadcast;
+use tokio_postgres::Client;
+
+use crate::election::{
+    Election, LeaderChangeMessage, CANDIDATES_ROOT, CANDIDATE_LEASE_SECS, ELECTION_KEY,
+};
+use crate::error::{
+    DeserializeFromJsonSnafu, PostgresExecutionSnafu, Result, SerializeToJsonSnafu, UnexpectedSnafu,
+};
+use crate::metasrv::{ElectionRef, LeaderValue, MetasrvNodeInfo};
+
+/// Value with a expire time. The expire time is in seconds since UNIX epoch.
+#[derive(Debug, Serialize, Deserialize, Default)]
+struct ValueWithLease {
+    value: String,
+    expire_time: f64,
+}
+
+/// PostgreSql implementation of Election.
+/// TODO: Currently only support candidate registration. Add election logic.
+pub struct PgElection {
+    leader_value: String,
+    client: Client,
+    is_leader: AtomicBool,
+    infancy: AtomicBool,
+    leader_watcher: broadcast::Sender<LeaderChangeMessage>,
+    store_key_prefix: String,
+    candidate_lease_ttl_secs: u64,
+}
+
+impl PgElection {
+    pub async fn with_pg_client(
+        leader_value: String,
+        client: Client,
+        store_key_prefix: String,
+        candidate_lease_ttl_secs: u64,
+    ) -> Result<ElectionRef> {
+        let (tx, _) = broadcast::channel(100);
+        Ok(Arc::new(Self {
+            leader_value,
+            client,
+            is_leader: AtomicBool::new(false),
+            infancy: AtomicBool::new(true),
+            leader_watcher: tx,
+            store_key_prefix,
+            candidate_lease_ttl_secs,
+        }))
+    }
+
+    fn _election_key(&self) -> String {
+        format!("{}{}", self.store_key_prefix, ELECTION_KEY)
+    }
+
+    fn candidate_root(&self) -> String {
+        format!("{}{}", self.store_key_prefix, CANDIDATES_ROOT)
+    }
+
+    fn candidate_key(&self) -> String {
+        format!("{}{}", self.candidate_root(), self.leader_value)
+    }
+}
+
+#[async_trait::async_trait]
+impl Election for PgElection {
+    type Leader = LeaderValue;
+
+    fn subscribe_leader_change(&self) -> broadcast::Receiver<LeaderChangeMessage> {
+        self.leader_watcher.subscribe()
+    }
+
+    fn is_leader(&self) -> bool {
+        self.is_leader.load(Ordering::Relaxed)
+    }
+
+    fn in_infancy(&self) -> bool {
+        self.infancy
+            .compare_exchange(true, false, Ordering::Relaxed, Ordering::Relaxed)
+            .is_ok()
+    }
+
+    async fn register_candidate(&self, node_info: &MetasrvNodeInfo) -> Result<()> {
+        let key = self.candidate_key();
+        let node_info =
+            serde_json::to_string(node_info).with_context(|_| SerializeToJsonSnafu {
+                input: format!("{node_info:?}"),
+            })?;
+        let value_with_lease = ValueWithLease {
+            value: node_info,
+            expire_time: time::SystemTime::now()
```

Comment:
Can we use the database server's time? I'm concerned that using client-side time might be inaccurate.